Port WebExtension Resources to C++

ItsJamie9494 opened the pull request:
#### 817155187498b088c6c2fb7a32f9b022245e047f
<pre>
Port WebExtension Resources to C++
<a href="https://webkit.org/b/281036">https://webkit.org/b/281036</a>

Reviewed by Timothy Hatcher.

Create a HashMap that stores raw bytes (span&lt;uint8_t&gt;) for resources, and use that instead of NSData.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toDataMap):
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension _initWithManifestDictionary:resources:]):
(-[WKWebExtension _initWithResources:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::resourceTypeForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::resourceFileURLForPath): Deleted.
(WebKit::WebExtension::resourceStringForPath): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::resourceFileURLForPath):
(WebKit::WebExtension::resourceStringForPath):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):

Canonical link: <a href="https://commits.webkit.org/285082@main">https://commits.webkit.org/285082@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/143962df1e0f18564a8d5108350c1bb63c1af3d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71492 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75603 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22697 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73607 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22516 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56462 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14933 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74558 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46193 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61584 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36911 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42856 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19048 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21038 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77322 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15726 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18587 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64174 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15769 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61623 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64170 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12320 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5958 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10956 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46705 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1484 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47776 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49060 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47518 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -128,18 +128,18 @@
-    RELEASE_ASSERT(m_resourceBaseURL.get().isFileURL);
-    RELEASE_ASSERT(m_resourceBaseURL.get().hasDirectoryPath);
+    RELEASE_ASSERT(m_resourceBaseURL.protocolIsFile());
+    RELEASE_ASSERT(m_resourceBaseURL.hasPath());
```
This likely does not translate, since the previous assert was basically looking for a trailing `/` on the path.

ItsJamie9494 replied:
I can add another check to see if the last character is a '/'

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +317,94 @@
-        if (auto base64Range = [path rangeOfString:@";base64,"]; base64Range.location != NSNotFound) {
-            auto *base64String = [path substringFromIndex:NSMaxRange(base64Range)];
-            return [[NSData alloc] initWithBase64EncodedString:base64String options:0];
+    NSString *cocoaPath = static_cast<NSString *>(path);
```
```suggestion
    auto *cocoaPath = static_cast<NSString *>(path);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -742,7 +692,9 @@
 
-    NSData *imageData = resourceDataForPath(imagePath, outError);
+    auto data = resourceDataForPath(imagePath, outError);
+    NSData *imageData = static_cast<NSData *>(API::Data::create(data)->wrapper());
```
```suggestion
    auto *imageData = static_cast<NSData *>(API::Data::create(data)->wrapper());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,65 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    auto resourceURL = URL(m_resourceBaseURL, path);
+
+    // Don't allow escaping the base URL with "../".
+    if (!resourceURL.string().startsWith(m_resourceBaseURL.string())) {
```
I'm skeptical that this does the same thing as before. The NSURL path resolved all the `../` and `./` in the path before doing this prefix check. Is that being done here?

xeenon replied:
Specifically the `URLByStandardizingPath` part of the old code. Not doing that would cause a security issue here.

ItsJamie9494 replied:
Looking at it again, along with the definitions for WTF::Url, I'm not sure if there's a way for WTF::Url to do that (and also, I'd imagine path resolving acts different on different platforms). Unless there's something in WTF I'm missing, I wonder if it would be better to take an NSUrl and get the URLByStandardisingPath on Cocoa, and do something different on other platforms

ItsJamie9494 replied:
Actually, I may have missed something. There is a `fileSystemPath()` function in URL, I wonder if that would resolve first?

xeenon replied:
I think `fileSystemPath()` might still have the `../`. You should test to verify if it does or not, an automation test would be good here too. I would be fine keeping a Cocoa only path for this for now, but it is a pretty vulnerable hole to have on other platforms.

ItsJamie9494 replied:
Yeah that's fine, I can investigate doing something different for glib later

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,65 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        return generatedBackgroundContent();
+
+    if (isServiceWorker && (path == generatedBackgroundServiceWorkerFilename))
+        return generatedBackgroundContent();
```
This was simplified in an earlier change. Let's keep the simple version.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,65 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    auto data = resourceDataForPath(path, outError, CacheResult::No, suppressErrors);
+    auto resourceString = emptyString();
+    if (!data.empty())
+        resourceString = String::fromUTF8(data);
```
We can't assume UTF8 here. The previous code guessed the string encoding. WTF should have something similar.

ItsJamie9494 replied:
Unless I'm missing something, it doesn't look like there's a way to guess with WTF::String. There's a function that falls back to Latin-1 though, which could be better here, but I don't know if another encoding is required.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,65 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        return { };
+
+    if (cacheResult == CacheResult::Yes)
+        m_resources.set(path, resourceString.span8());
```
Is `span8()` safe for Unicode? Uses UTF8?

You used `utf8().span()` elsewhere.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -165,7 +165,7 @@
 bool WebExtensionAPIPermissions::verifyRequestedPermissions(HashSet<String>& permissions, HashSet<Ref<WebExtensionMatchPattern>>& matchPatterns, NSString *callingAPIName, NSString **outExceptionString)
 {
-    auto extension = WebExtension::create(extensionContext().manifest(), @{ });
+    auto extension = WebExtension::create(extensionContext().manifest(), HashMap<String, std::span<const uint8_t>>());
```
```suggestion
    auto extension = WebExtension::create(extensionContext().manifest(), { });
```

ItsJamie9494 replied:
`error: no matching function for call to 'create'`.

xeenon replied:
Can you change `explicit WebExtension(NSDictionary *manifest, const Resources&)` to be `explicit WebExtension(NSDictionary *manifest, const Resources& = { })`, and just remove the empty object at this call site then?

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+        auto *keyString = dynamic_objc_cast<NSString>(key);
+        if (!keyString)
+            continue;
+        id value = dictionary[key];
```
```suggestion
            continue;

        id value = dictionary[key];
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+            result.add(keyString, dataSpan);
+            continue;
+        }
+        if (auto *valueData = dynamic_objc_cast<NSData>(value)) {
```
```suggestion
        }

        if (auto *valueData = dynamic_objc_cast<NSData>(value)) {
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+            result.add(keyString, dataSpan);
+            continue;
+        }
+        if ([value isKindOfClass:[NSDictionary class]] || [value isKindOfClass:[NSArray class]]) {
```
```suggestion
        }

        if ([value isKindOfClass:[NSDictionary class]] || [value isKindOfClass:[NSArray class]]) {
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+            continue;
+        }
+    }
+    return result;
```
```suggestion
    }

    return result;
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+            auto *jsonData = encodeJSONData(value, JSONOptions::FragmentsAllowed, &error);
+            if (!jsonData || error)
+                continue;
+            dataSpan = std::span<const uint8_t>(reinterpret_cast<const uint8_t*>(jsonData.bytes), jsonData.length);
```
```suggestion
                continue;

            dataSpan = std::span<const uint8_t>(reinterpret_cast<const uint8_t*>(jsonData.bytes), jsonData.length);
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,37 @@ WallTime toImpl(NSDate *date)
+{
+    HashMap<String, std::span<const uint8_t>> result;
+    result.reserveInitialCapacity(dictionary.count);
+    for (id key in dictionary) {
```
```suggestion
    result.reserveInitialCapacity(dictionary.count);

    for (id key in dictionary) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        return { };
+    }
+
+    return URL(resourceURL);
```
This will not be a `file` URL. I suspect this will fail to parse, since `resourceURL` will just be a `/foo/bar/path` string.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    return URL(resourceURL);
+}
+
+String WebExtension::resourceStringForPath(const String& unsafePath, RefPtr<API::Error>& outError, CacheResult cacheResult, SuppressNotFoundErrors suppressErrors)
```
I like `originalPath` better.
```suggestion
String WebExtension::resourceStringForPath(const String& originalPath, RefPtr<API::Error>& outError, CacheResult cacheResult, SuppressNotFoundErrors suppressErrors)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +245,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    auto data = resourceDataForPath(path, outError, CacheResult::No, suppressErrors);
+    auto resourceString = emptyString();
+    if (!data.empty())
+        resourceString = String(data);
```
This constructor will use latin-1, so this is worse. Use `String::fromUTF8WithLatin1Fallback()`. It is still not as robust as the Cocoa implementation, with would support any text encoding (UTF-16, Shift JIS, etc.)

ItsJamie9494 replied:
Yeah, I know it's not as robust but unless I'm missing something WTF doesn't have support for others

xeenon replied:
We have `TextResourceDecoder::create()` that will do this. We should use that here.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +247,68 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        return { };
+
+#if PLATFORM(COCOA)
+    auto resourceURL = static_cast<NSURL *>(URL(m_resourceBaseURL, path)).URLByStandardizingPath;
```
This will need to be in the platform specific file as it is ObjC.

Also `resourceURL` should be renamed to `resourcePath`.

ItsJamie9494 commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +248,67 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
         parseWebAccessibleResourcesVersion2();
 }
 
+#if !PLATFORM(COCOA)
```
I may remove this, as unless I can figure out and test a way (with WTF and WebCore) to expand a path string, it'll have to stay platform specific, as I'd prefer not to introduce vulnerabilities :)

xeenon replied:
I looked into it, you can use `FileSystem::realPath()` to do this in a cross-platform way.

From the `man` page:
```
The realpath() function resolves all symbolic links, extra “/” characters, and
references to /./ and /../ in file_name.
```

ItsJamie9494 commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
```diff
@@ -107,13 +109,15 @@
-            extensionContext->recordError(wrapper(error));
+        auto resourceData = extensionContext->extension().resourceDataForPath(requestURL.path().toString(), error);
+        if (resourceData.empty()) {
+            extensionContext->recordErrorIfNeeded(wrapper(error));
```
On my local machine, this failed as recordError because no error was being passed. Using recordErrorAsNeeded lets it compile, but doesn't pass tests (they all timeout at 33 seconds locally).

xeenon replied:
I'm not sure. I can try to build your branch locally sometime tomorrow and look into it.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,43 @@ WallTime toImpl(NSDate *date)
+        }
+
+        if (auto *valueData = dynamic_objc_cast<NSData>(value)) {
+            dataSpan = std::span<const uint8_t>(reinterpret_cast<const uint8_t*>(valueData.bytes), valueData.length);
```
```suggestion
            dataSpan = API::Data::createWithoutCopying(valueData).span();
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -537,4 +537,43 @@ WallTime toImpl(NSDate *date)
+            if (!jsonData || error)
+                continue;
+
+            dataSpan = std::span<const uint8_t>(reinterpret_cast<const uint8_t*>(jsonData.bytes), jsonData.length);
```
```suggestion
            dataSpan = API::Data::createWithoutCopying(jsonData).span();
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
```diff
@@ -154,7 +154,8 @@ - (instancetype)_initWithManifestDictionary:(NSDictionary<NSString *, id> *)mani
 
-    API::Object::constructInWrapper<WebKit::WebExtension>(self, manifest, resources);
+    auto resourcesData = WebKit::toDataMap(resources);
+    API::Object::constructInWrapper<WebKit::WebExtension>(self, manifest, WTFMove(resourcesData));
```
```suggestion
    API::Object::constructInWrapper<WebKit::WebExtension>(self, manifest, WebKit::toDataMap(resources));
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
```diff
@@ -166,7 +167,8 @@ - (instancetype)_initWithResources:(NSDictionary<NSString *, id> *)resources
 
-    API::Object::constructInWrapper<WebKit::WebExtension>(self, resources);
+    auto resourcesData = WebKit::toDataMap(resources);
+    API::Object::constructInWrapper<WebKit::WebExtension>(self, WTFMove(resourcesData));
```
```suggestion
    API::Object::constructInWrapper<WebKit::WebExtension>(self, WebKit::toDataMap(resources));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -149,21 +150,21 @@
 }
 
-WebExtension::WebExtension(NSDictionary *manifest, NSDictionary *resources)
+WebExtension::WebExtension(NSDictionary *manifest, const Resources& resources)
```
```suggestion
WebExtension::WebExtension(NSDictionary *manifest, Resources&& resources)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -149,21 +150,21 @@
+WebExtension::WebExtension(NSDictionary *manifest, const Resources& resources)
     : m_manifestJSON(JSON::Value::null())
-    , m_resources([resources mutableCopy] ?: [NSMutableDictionary dictionary])
+    , m_resources(resources)
```
```suggestion
    , m_resources(WTFMove(resources))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -149,21 +150,21 @@
 }
 
-WebExtension::WebExtension(NSDictionary *resources)
+WebExtension::WebExtension(const Resources& resources)
```
```suggestion
WebExtension::WebExtension(Resources&& resources)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -149,21 +150,21 @@
+WebExtension::WebExtension(const Resources& resources)
     : m_manifestJSON(JSON::Value::null())
-    , m_resources([resources mutableCopy] ?: [NSMutableDictionary dictionary])
+    , m_resources(resources)
```
```suggestion
    , m_resources(WTFMove(resources))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
+    if ([cocoaPath hasPrefix:@"data:"]) {
+        if (auto base64Range = [cocoaPath rangeOfString:@";base64,"]; base64Range.location != NSNotFound) {
+            auto *base64String = [cocoaPath substringFromIndex:NSMaxRange(base64Range)];
+            auto data = retainPtr([[NSData alloc] initWithBase64EncodedString:base64String options:0]);
```
No need for `RetainPtr` here, since we use ObjC ARC.
```suggestion
            auto *data = [[NSData alloc] initWithBase64EncodedString:base64String options:0];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
+            auto *urlEncodedString = [cocoaPath substringFromIndex:NSMaxRange(commaRange)];
             auto *decodedString = [urlEncodedString stringByRemovingPercentEncoding];
-            return [decodedString dataUsingEncoding:NSUTF8StringEncoding];
+            auto data = retainPtr([decodedString dataUsingEncoding:NSUTF8StringEncoding]);
```
```suggestion
            auto *data = [decodedString dataUsingEncoding:NSUTF8StringEncoding];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
-        ASSERT([path isEqualToString:@"data:"]);
-        return [NSData data];
+        ASSERT([cocoaPath isEqualToString:@"data:"]);
+        return API::Data::createWithoutCopying([NSData data])->span();
```
```suggestion
        return { };
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
 
-        return result;
+        return *cachedObject;
     }
```
This is doing a lot of work just to get back to the same data result that is in the cache.
```suggestion
    if (auto cachedObject = m_resources.getOptional(path))
        return *cachedObject;
````

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
     }
 #endif
-
+    auto data = API::Data::createWithoutCopying(retainPtr(resultData))->span();
```
```suggestion
    auto data = API::Data::createWithoutCopying(resultData)->span();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -339,122 +318,117 @@
+    }
+
+    return resourceURL;
 }
```
Can you add this back where it was, above `WebExtension::resourceTypeForPath`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4352,13 +4354,13 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
             }
 
-            auto userScript = API::UserScript::create(WebCore::UserScript { scriptString, URL { m_baseURL, scriptPath }, makeVector<String>(includeMatchPatterns), makeVector<String>(excludeMatchPatterns), injectionTime, injectedFrames, waitForNotification }, executionWorld);
+            auto userScript = API::UserScript::create(WebCore::UserScript { WTFMove(scriptString), URL { m_baseURL, scriptPath }, makeVector<String>(includeMatchPatterns), makeVector<String>(excludeMatchPatterns), injectionTime, injectedFrames, waitForNotification }, executionWorld);
```
`String` does not need a move, since it is a thin wrapper around `StringImpl`.

```suggestion
            auto userScript = API::UserScript::create(WebCore::UserScript { scriptString, URL { m_baseURL, scriptPath }, makeVector<String>(includeMatchPatterns), makeVector<String>(excludeMatchPatterns), injectionTime, injectedFrames, waitForNotification }, executionWorld);
```

ItsJamie9494 replied:
UserScript is a String&&, and without the move, the compiler complains about a missing rvalue. Generally, I've noticed other places that use a String&& move the string

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4645,13 +4647,13 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
-            auto *jsonData = extension->resourceDataForPath(ruleset.jsonPath, error);
-            if (!jsonData) {
+            auto jsonData = extension->resourceDataForPath(ruleset.jsonPath, error);
+            if (!jsonData.size()) {
```
```suggestion
            if (jsonData.empty()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4645,13 +4647,13 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
             }
 
-            [allJSONData addObject:jsonData];
+            [allJSONData addObject:(API::Data::create(jsonData)->wrapper())];
```
```suggestion
            [allJSONData addObject:API::Data::createWithoutCopying(jsonData)->wrapper()];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -214,14 +215,13 @@
     }
 
-    if (!parseManifest(manifestData))
+    if (!parseManifest(static_cast<NSData *>(API::Data::create(manifestData)->wrapper())))
```
```suggestion
    if (!parseManifest(static_cast<NSData *>(API::Data::createWithoutCopying(manifestData)->wrapper())))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -742,7 +716,9 @@
 
-    NSData *imageData = resourceDataForPath(imagePath, outError);
+    auto data = resourceDataForPath(imagePath, outError);
+    auto *imageData = static_cast<NSData *>(API::Data::create(data)->wrapper());
```
```suggestion
    auto *imageData = static_cast<NSData *>(API::Data::createWithoutCopying(data)->wrapper());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -85,13 +86,13 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
-    auto *scriptData = extensionContext.protectedExtension()->resourceDataForPath(path, error);
-    if (!scriptData) {
+    auto scriptData = extensionContext.protectedExtension()->resourceDataForPath(path, error);
+    if (!scriptData.size()) {
```
```suggestion
    if (scriptData.empty()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -85,13 +86,13 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
     }
 
-    return SourcePair { [[NSString alloc] initWithData:scriptData encoding:NSUTF8StringEncoding], { extensionContext.baseURL(), path } };
+    return SourcePair { [[NSString alloc] initWithData:wrapper(API::Data::create(scriptData)).get() encoding:NSUTF8StringEncoding], { extensionContext.baseURL(), path } };
```
```suggestion
    return SourcePair { String::fromUTF8(scriptData), { extensionContext.baseURL(), path } };
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
```diff
@@ -107,13 +109,15 @@
             return;
         }
 
+        auto fileData = wrapper(API::Data::create(resourceData)).get();
```
```suggestion
        auto *fileData = static_cast<NSData *>(API::Data::createWithoutCopying(resourceData)->wrapper());
```

ItsJamie9494 replied:
resourceData isn't an NSData, and createWithoutCopying only exists for std::span if we also pass in a freeDataFunction.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +247,67 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        path = path.substring(1);
+
+    if (auto cachedData = m_resources.getOptional(path)) {
+        if (auto cachedString = String::fromUTF8(*cachedData); !cachedString.isEmpty())
```
I think even if empty, just return it, since that is what was cached.
```suggestion
        if (auto cachedString = String::fromUTF8(*cachedData))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +247,67 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    if (!data.empty()) {
+        auto decoder = TextResourceDecoder::create("text/plain"_s);
+
+        resourceString = decoder->decode(data);
```
Need to pass `true` to use the auto detection.
```suggestion
        auto decoder = TextResourceDecoder::create("text/plain"_s, { }, true);
        resourceString = decoder->decode(data);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -69,15 +69,17 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
-    explicit WebExtension(NSDictionary *manifest, NSDictionary *resources);
-    explicit WebExtension(NSDictionary *resources);
+    explicit WebExtension(NSDictionary *manifest, const Resources& = { });
+    explicit WebExtension(const Resources& = { });
```
```suggestion
    explicit WebExtension(NSDictionary *manifest, Resources&& = { });
    explicit WebExtension(Resources&& = { });
```

xeenon reviewed (changes requested):
I took another full pass. This is getting close!

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -29,6 +29,8 @@
 #if ENABLE(WK_WEB_EXTENSIONS)
 
 #include "WebExtensionUtilities.h"
+#include <WebCore/CommonAtomStrings.h>
```
Is this needed?

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +247,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    if (!path.length() || m_resourceBaseURL.isNull())
+        return { };
+
+    auto resourceURL = FileSystem::realPath(URL(m_resourceBaseURL, path).fileSystemPath());
```
```suggestion
    auto resourcePath = FileSystem::realPath(URL(m_resourceBaseURL, path).fileSystemPath());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -245,6 +247,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+    auto resourceURL = FileSystem::realPath(URL(m_resourceBaseURL, path).fileSystemPath());
+
+    // Don't allow escaping the base URL with "../".
+    if (!resourceURL.startsWith(m_resourceBaseURL.string())) {
```
You need to get the real path to `m_resourceBaseURL.filesystemPath()` for this to work.

ItsJamie9494 committed:
Port WebExtension Resources to C++
https://webkit.org/b/281036

Reviewed by Timothy Hatcher.

Create a HashMap that stores raw bytes (span<uint8_t>) for resources, and use that instead of NSData.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toDataMap):
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension _initWithManifestDictionary:resources:]):
(-[WKWebExtension _initWithResources:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::resourceTypeForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::resourceFileURLForPath): Deleted.
(WebKit::WebExtension::resourceStringForPath): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::resourceFileURLForPath):
(WebKit::WebExtension::resourceStringForPath):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):

Canonical link: https://commits.webkit.org/285082@main
